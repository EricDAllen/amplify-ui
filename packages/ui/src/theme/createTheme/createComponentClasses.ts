interface ComponentClassesProps {
  modifier?: Array<string | undefined>;
  element?: Array<string | undefined>;
}

export function createComponentClasses({ name = '', prefix = 'amplify-' }) {
  return (props?: ComponentClassesProps) => {
    const baseComponentClassName = `${prefix}${name}`;
    if (!props) {
      return baseComponentClassName;
    }

    const className = Array.isArray(props.element)
      ? `${baseComponentClassName}__${props.element.join('__')}`
      : baseComponentClassName;
    const classNames = [className];

    if (Array.isArray(props.modifier)) {
      props.modifier.forEach((modifier) => {
        if (!modifier) {
          return;
        }
        classNames.push(`${className}--${modifier}`);
      });
    }

    return classNames.join(' ');
  };
}
