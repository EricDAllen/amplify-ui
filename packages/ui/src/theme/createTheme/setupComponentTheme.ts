import kebabCase from 'lodash/kebabCase.js';
import { WebTheme } from '../types';
import { cssValue } from './cssValue';
import { BaseComponentTheme, ComponentsTheme } from '../components';
import { isFunction } from '../../utils';
import { isDesignToken } from './isDesignToken';
import { CSSProperties } from '../components/utils';

// we need to handle psuedo elements like ::before

/**
 * This will take an object like:
 * {paddingTop:'20px',color:'{colors.font.primary}'}
 * and turn it into a CSS string:
 * `padding-top:20px; color: var(--colors-font-primary);`
 */
function propsToString(props: CSSProperties): string {
  return Object.entries(props)
    .map(([key, value]) => {
      const _value = isDesignToken(value)
        ? value.toString()
        : // @ts-ignore
          cssValue({ value });
      return `${kebabCase(key)}:${_value}; `;
    })
    .join(' ');
}

// function to split an object in 2 by a predicate on the keys
function splitObject(
  obj: Record<string, any>,
  predicate: (key: string) => boolean
) {
  const left = {};
  const right = {};
  Object.entries(obj).forEach(([key, value]) => {
    if (predicate(key)) {
      left[key] = value;
    } else {
      right[key] = value;
    }
  });
  return [left, right] as const;
}

function createComponentCSS(baseSelector: string, theme: BaseComponentTheme) {
  if (!theme) return '';
  let str = '';
  const { modifier = {}, element = {}, vars, ...props } = theme;

  // if there are no props, skip
  if (Object.keys(props).length) {
    // separate psuedo/attribute selectors
    const [selectors, other] = splitObject(
      props,
      (key) => key.startsWith(':') || key.startsWith('[')
    );

    Object.entries(selectors).forEach(([key, value]) => {
      str += `${baseSelector}${key} { ${propsToString(value)} }\n`;
    });

    str += `${baseSelector} { ${propsToString(other)} }\n`;
  }

  if (vars) {
    Object.entries(vars).forEach(([key, value]) => {
      str += `${baseSelector} { --${key}:${value}; }\n`;
    });
  }

  Object.entries(modifier).forEach(([key, value]) => {
    if (value && Object.keys(value).length) {
      str += createComponentCSS(`${baseSelector}--${key}`, value);
    }
  });

  Object.entries(element).forEach(([key, value]) => {
    if (value && Object.keys(value).length) {
      str += createComponentCSS(`${baseSelector}__${key}`, value);
    }
  });

  return str;
}

export function setupComponentTheme(
  str: string,
  components: ComponentsTheme,
  tokens: WebTheme['tokens']
) {
  let cssText = '';

  Object.entries(components).forEach(([key, component]) => {
    const baseComponentClassName = `amplify-${key}`;
    const componentClassName = `${str} .${baseComponentClassName}`;
    // unwrap the component theme
    // if it is a function: call it with the defaultTheme to get a static object
    const componentTheme: BaseComponentTheme = isFunction(component)
      ? (component(tokens) as BaseComponentTheme)
      : component;

    cssText += createComponentCSS(componentClassName, componentTheme);
  });

  return cssText;
}
