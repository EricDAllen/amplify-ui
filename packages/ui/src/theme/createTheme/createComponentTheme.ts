import { BaseComponentTheme, ComponentTheme } from '../components';
import { createComponentClasses } from './createComponentClasses';

type CreactComponentThemeProps<ThemeType> = {
  name?: string;
  prefix?: string;
  theme: ThemeType;
};

export function createComponentTheme<
  ThemeType extends
    ComponentTheme<BaseComponentTheme> = ComponentTheme<BaseComponentTheme>,
>({
  name = '',
  prefix = 'amplify-',
  theme,
}: CreactComponentThemeProps<ThemeType>): {
  className: (props?: { modifier?: string[]; element?: string[] }) => string;
  theme: ThemeType;
} {
  const className = createComponentClasses({ name, prefix });
  return {
    className,
    theme,
  };
}
