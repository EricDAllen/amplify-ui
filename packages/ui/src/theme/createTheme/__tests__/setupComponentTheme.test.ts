import { setupComponentTheme } from '../setupComponentTheme';
import { createTheme } from '../createTheme';

const { tokens } = createTheme();

describe('@aws-amplify/ui', () => {
  describe('setupComponentTheme', () => {
    it('should pass through raw values', () => {
      const css = setupComponentTheme(
        `[data-amplify-theme="test"]`,
        {
          badge: {
            backgroundColor: 'pink',
            borderRadius: '{radii.small}',
            modifier: {
              small: {
                borderRadius: '0',
              },
            },
          },
        },
        tokens
      );
      expect(css).toMatchSnapshot();
    });

    it('should button', () => {
      expect(
        setupComponentTheme(
          `[data-amplify-theme="test"]`,
          {
            button: {
              modifier: {
                primary: {
                  background: 'red',
                  ':hover': {
                    background: 'pink',
                    color: '{colors.font.primary}',
                  },
                },
              },
            },
          },
          tokens
        )
      ).toMatchSnapshot();
    });

    it('can use custom primitives', () => {
      const css = setupComponentTheme(
        `[data-amplify-theme="test"]`,
        {
          chip: {
            paddingInline: '{space.xs}',
            paddingBlock: '{space.xs}',
            borderRadius: '{radii.small}',
            modifier: {
              primary: {
                background: 'red',
                ':hover': {
                  background: 'pink',
                  color: '{colors.font.primary}',
                },
              },
            },
          },
        },
        createTheme().tokens
      );
      expect(css).toMatchSnapshot();
    });
  });
});
