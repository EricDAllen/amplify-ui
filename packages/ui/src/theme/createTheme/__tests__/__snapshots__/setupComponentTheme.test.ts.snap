// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@aws-amplify/ui setupComponentTheme can use custom primitives 1`] = `
"[data-amplify-theme="test"] .amplify-chip { padding-inline:var(--amplify-space-xs);  padding-block:var(--amplify-space-xs);  border-radius:var(--amplify-radii-small);  }
[data-amplify-theme="test"] .amplify-chip--primary:hover { background:pink;  color:var(--amplify-colors-font-primary);  }
[data-amplify-theme="test"] .amplify-chip--primary { background:red;  }
"
`;

exports[`@aws-amplify/ui setupComponentTheme should button 1`] = `
"[data-amplify-theme="test"] .amplify-button--primary:hover { background:pink;  color:var(--amplify-colors-font-primary);  }
[data-amplify-theme="test"] .amplify-button--primary { background:red;  }
"
`;

exports[`@aws-amplify/ui setupComponentTheme should pass through raw values 1`] = `
"[data-amplify-theme="test"] .amplify-badge { background-color:pink;  border-radius:var(--amplify-radii-small);  }
[data-amplify-theme="test"] .amplify-badge--small { border-radius:0;  }
"
`;
